d3-series/materials/Week 6/Assignments/us-station.html

```diff
@@ -0,0 +1,174 @@
+                    handleError(error, 'failed to read geoJSON data');
+
+                    d3.csv('NSRDB_StationsMeta.csv', function(error, stations) {
+                        handleError(error, 'failed to read station data');
+                        stations=prepDataStations(stations);
+                        console.log(stations);
+                        draw(geojson, stations);
+                        
+                        
+                    });
+                });
+
+                function handleError(error, msg) {
+                    if (error) {
+                        console.error(msg);
+                    }
+                }
+
+                function prepDataStations(data) {
+                    return data
+                    .map(function(d) {
+                        return {
+                            name: d.USAF,
+                            class: +d.CLASS,
+                            loc: [+d.longitude, +d.latitude],
+                            ele: +d.NSRDB_ELEV
+                        };
+                    });
+                }
+                function draw(geojson, stations) {
+
+                    var colors = ['red','yellow','blue'];
+
+                    var scale = d3
+                                .scaleLinear()
+                                .domain([
+                                    d3.min(stations, function(d) {
+                                        return d.ele;
+                                    }),
+                                    d3.max(stations, function(d) {
+                                        return d.ele;
+                                    })
+                                ])
+                                .range([2,15]);
+
+                    var usaProj = d3
+                        .geoAlbersUsa()
+                        //.scale(1000)
+                        .translate([innerWidth / 2, innerHeight / 2]);
+
+                    var geoPath = d3.geoPath().projection(usaProj);
+
+                    g
+                        .selectAll('path')
+                        .data(geojson.features)
+                        .enter()
+                        .append('path')
+                        .attr('d', geoPath)
+                        .style('fill', 'white')
+                        .style('stroke', 'black')
+                        .style('stroke-width', 0.5);
+
+                    g
+                        .selectAll('circle')
+                        .data(stations)
+                        .enter()
+                        .append('circle')
+                        .attr('cx', function(d) {
+                            if(usaProj(d.loc)!=null){
+                                return usaProj(d.loc)[0];
+                            }
+                            
+                        })
+                        .attr('cy', function(d) {
+                            if(usaProj(d.loc)!=null){

```

Comment:
To avoid repeating the logic for filtering out the null location values with the conditional for the cy/cx properties, the null values could be filtered out beforehand and the filtered data could be passed in as the bound data for the circle elements.